python/pyarrow/tests/test_flight.py

```diff
@@ -862,3 +862,37 @@ def test_roundtrip_errors():
             list(client.do_action(flight.Action("unauthorized", b"")))
         with pytest.raises(flight.FlightInternalError, match=".*foo.*"):
             list(client.list_flights())
+
+
+def test_do_put_independent_read_write():
+    """Ensure that separate threads can read/write on a DoPut."""
+    # ARROW-6063: previously this would segfault when the writer was
+    # closed, or would hang forever.
```

Comment:
Hmm... why segfault? Isn't this something that we can detect and prevent programmatically? (raise an error if reading from a stream whose writer side was closed)